src/librbd/ImageWatcher.cc

```diff
@@ -130,42 +56,7 @@ void ImageWatcher<I>::unregister_watch(Context *on_finish) {
 
   cancel_async_requests();
 
-  C_Gather *gather_ctx = nullptr;
-  {
-    RWLock::WLocker watch_locker(m_watch_lock);
-    if (m_watch_state == WATCH_STATE_REWATCHING) {
-      ldout(cct, 10) << this << " delaying unregister until rewatch completed"
-                     << dendl;
-
-      assert(m_unregister_watch_ctx == nullptr);
-      m_unregister_watch_ctx = new FunctionContext([this, on_finish](int r) {
-          unregister_watch(on_finish);
-        });
-      return;
-    }
-
-    gather_ctx = new C_Gather(m_image_ctx.cct, create_async_context_callback(
-      m_image_ctx, on_finish));
-    if (m_watch_state == WATCH_STATE_REGISTERED ||
-        m_watch_state == WATCH_STATE_ERROR) {
-      m_watch_state = WATCH_STATE_UNREGISTERED;
-
-      librados::AioCompletion *aio_comp = create_rados_safe_callback(
-        new C_UnwatchAndFlush(m_image_ctx.md_ctx, gather_ctx->new_sub()));
-      int r = m_image_ctx.md_ctx.aio_unwatch(m_watch_handle, aio_comp);
-      assert(r == 0);
-      aio_comp->release();
-    }
-  }
-
-  assert(gather_ctx != nullptr);
-  m_task_finisher->cancel_all(gather_ctx->new_sub());
```

Comment:
Almost all logic was moved to the `Watcher::unregister_watch`, although I forgot to call the `cancel_all` method in the subclass.